sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -374,8 +375,12 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
     df.sparkSession.sessionState.sqlParser.parseMultipartIdentifier(tableName) match {
       case CatalogObjectIdentifier(Some(catalog), ident) =>
         insertInto(catalog, ident)
+      // TODO(SPARK-28667): Support the V2SessionCatalog
```

Comment:
Not related to this PR, but just a note. I've been working on V2SessionCatalog improvement for a while, and one issue I found is: it's not easy to implement "if the table provider is v1, go to v1 session catalog. if the table provider is v2, go to v2 session catalog".

It's easy to do it for CREATE TABLE, because the table provider is known at the beginning. But it's hard for SELECT and INSERT, as we need to look up the table from Hive catalog to get the table provider.

I'd expect to have 2 analyzer rules to do it:
1. one rule in sql/catalyst, which resolves `UnresolvedRelation` to `UnresolvedCatalogRelation` by looking up the table from hive catalog
2. one rule in sql/core, which resolves `UnresolvedCatalogRelation` to `DataSourceV2Relation`, if the table provider is v2. This has to be in sql/core because `DataSource.lookupDataSource` is in sql/core. We have a rule `FindDataSourceTable` that resolves `UnresolvedCatalogRelation` to v1 relation if the table provider is v1.

I think `DataFrameWriter` should produce `...Statement` plans, to reuse the analyzer rules for all the things.

Reply:
Well, the V2SessionCatalog already produces a `CatalogTableAsV2`, which gets unwrapped to the `UnresolvedCatalogRelation` already in `DataSourceResolution`, and then it can get resolved to the V1 code paths. I've also had a similar problem when trying to implement V2SessionCatalog support for `ALTER TABLE`. In that case I decided that it should reject and fallback to V1 plans if the loaded table is a `CatalogTableAsV2`.

Reply:
> V2SessionCatalog already produces a CatalogTableAsV2

Yes, for both v1 and v2 provider. So we still need a rule to check the table provider, and either unwrap it to a `UnresolvedCatalogRelation`, or create the actual v2 table.

Reply:
> Yes, for both v1 and v2 provider. So we still need a rule to check the table provider, and either unwrap it to a UnresolvedCatalogRelation, or create the actual v2 table.

This should be done in `loadTable`. The original implementation called `TableProvider` to create the table, but this path is currently broken. I think that `TableProvider` needs to be improved so that it works instead of adding more rules to convert from `CatalogTableAsV2`. 

The catalog should return the correct `Table` instance for all v2 tables. Spark shouldn't convert between v2 table instances.